include/fixed_containers/optional_ref.hpp

```diff
@@ -0,0 +1,192 @@
+#pragma once
+
+#include "fixed_containers/concepts.hpp"
+#include "fixed_containers/optional_ref_checking.hpp"
+#include "fixed_containers/source_location.hpp"
+
+#include <compare>
+#include <functional>
+#include <optional>
+
+namespace fixed_containers
+{
+/*
+ * OptionalRef<T> is intended to mimic std::optional<std::reference_wrapper<T>>
+ * but without the extra indirection when dereferencing.
+ */
+template <IsNotReference T,
+          customize::OptionalRefChecking<T> CheckingType = customize::OptionalRefAbortChecking<T>>
+class OptionalRef
+{
+private:
+    using Self = OptionalRef<T, CheckingType>;
+    using reference = T&;
+    using const_reference = const T&;
+    using pointer = T*;
+    using const_pointer = const T*;
+    using BackingType = pointer;
+
+public:
+    // Needed for structural type
+    BackingType IMPLEMENTION_DETAIL_DO_NOT_USE_underlying_val_ = nullptr;
+
+public:
+    constexpr OptionalRef() = default;
+    constexpr OptionalRef(T& val)
+      : IMPLEMENTION_DETAIL_DO_NOT_USE_underlying_val_(&val)
+    {
+    }
+
+    constexpr ~OptionalRef() = default;
+
+    constexpr OptionalRef(const Self&) noexcept = default;
+    constexpr Self& operator=(const Self&) = delete;  // copy assigment is ambiguous for this type
+
+    constexpr OptionalRef(Self&&) noexcept = default;
+    constexpr Self& operator=(Self&&) = default;
```

Comment:
thoughts about the move assignment operator being ambiguous for this type?
ie. 
```cpp
int a = 5
OptionalRef opt(a);
int b = 10;
OptionalRef opt2(b);
opt2 = std::move(opt)
```

currently b stays at 10, and opt2 just rebinds. IMO this makes intuitive sense

Reply:
Some relevant info;
- https://www.boost.org/doc/libs/1_85_0/libs/optional/doc/html/boost_optional/tutorial/optional_references.html
- https://stackoverflow.com/questions/26858034/why-does-stdoptional-not-have-a-specialization-for-reference-types

Reply:
sync'd offline: We will allow assignment operators for `std::nullopt_t` and `OptionalRef<T>` but not `T`. This is motivated from this proposal
https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2012/n3406#rationale.refs

```cpp
int i = 0;
int j = 1;
optional<int&> ori;
optional<int&> orj = j;

*orj = 2;
assert (j == 2);

ori = i; // ERROR: no assignment from int&
ori = {i}; // OK: assignemnt from optional<int&>

orj = ori; // OK: rebinding assignemnt from optional<int&>
*orj = 4;
assert (j == 2);
assert (i == 4);
```